### DISCUSSION TITLE
TX2 NX tach_gpio

### DISCUSSION THREAD

**Discussion by dat-tran-matternet** [OE4T/meta-tegra#958] [General]
I noticed that the TX2 NX device tree didn't have tach_gpio defined. Here is the patch if anyone is interested. You can enable and read afterwards with:

```
echo 1 > /sys/devices/pwm-fan/tach_enable
cat /sys/devices/pwm-fan/rpm_measured
```

```
diff --git a/nvidia/platform/t18x/lanai/kernel-dts/common/tegra186-p3636-0001-a00-pwm-fan.dtsi b/nvidia/platform/t18x/lanai/kernel-dts/common/tegra186-p3636-0001-a00-pwm-fan.dtsi
index 769be8c97bb3..7a422d774d37 100644
--- a/nvidia/platform/t18x/lanai/kernel-dts/common/tegra186-p3636-0001-a00-pwm-fan.dtsi
+++ b/nvidia/platform/t18x/lanai/kernel-dts/common/tegra186-p3636-0001-a00-pwm-fan.dtsi
@@ -30,6 +30,7 @@
 		state_cap = <7>;
 		active_pwm_max = <256>;
 		tach_period = <1000>;
+		tach_gpio = <&tegra_main_gpio TEGRA_MAIN_GPIO(X, 4) GPIO_ACTIVE_LOW>;
 		pwm_gpio = <&tegra_aon_gpio TEGRA_AON_GPIO(V, 6) GPIO_ACTIVE_LOW>; /* TEGRA_MAIN_GPIO_PV6 */
 		pwm_polarity = <PWM_POLARITY_NORMAL>;
 		suspend_state = <0>;
-- 

```

**Comment by madisongh** [comment]
With recent L4T versions, NVIDIA switched to using a `generic-pwm-tachometer` driver that creates a node under `/sys/class/hwmon` where you can retrieve the fan RPM.  See [this post in NVIDIA's dev forum](https://forums.developer.nvidia.com/t/tach-gpio-is-missing-from-pwm-fan-device-tree-so-we-cant-monitor-fan-rpm/128331/8).

**Reply by dat-tran-matternet** [reply]
Would that method give the pwm fan controller the feedback it needs? Assuming it's a closed loop controller. Thanks

**Reply by madisongh** [reply]
I'm don't the cooling driver itself uses the tachometer reading, but just sets a desired RPM based on current temperature.  You can check the relevant drivers, though. I use the tach reading on a TX2-based project for testing that my fan curve table is correct (custom carrier with custom fan). It's been a while since I looked at it, but IIRC the new driver makes use of a hardware block on the chip for the tach vs. the software sampling approach that the old pwm-fan driver used.
